pkg/services/ngalert/store/alertmanager.go

```diff
@@ -132,6 +143,59 @@ func (st *DBstore) UpdateAlertmanagerConfiguration(ctx context.Context, cmd *mod
 	})
 }
 
+// MarkConfigurationAsApplied sets the `applied_at` field of the last config with the given hash to the current UNIX timestamp.
+func (st *DBstore) MarkConfigurationAsApplied(ctx context.Context, cmd *models.MarkConfigurationAsAppliedCmd) error {
+	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *db.Session) error {
+		// Create a map representing the field to be updated and its new value.
+		update := map[string]interface{}{"applied_at": time.Now().UTC().Unix()}
+
+		// Set `applied_at` in the configurations table.
+		rowsAffected, err := sess.Table("alert_configuration").Where("id = ?", cmd.ConfigID).Update(&update)
+		if err != nil {
+			return err
+		}
+
+		if rowsAffected != 1 {
+			return fmt.Errorf("update statement affected %d alert configuration records", rowsAffected)
+		}
+
+		// Update the same field in the config history table.
+		rowsAffected, err = sess.Table("alert_configuration_history").
+			Desc("id").
+			Limit(1).
+			Where("org_id = ? AND configuration_hash = ?", cmd.OrgID, cmd.ConfigurationHash).
+			Cols("applied_at").
+			Update(&update)
+
+		if err != nil {
+			return err
+		}
+
+		if rowsAffected != 1 {
+			return fmt.Errorf("update statement affected %d alert configuration history records", rowsAffected)
+		}
+		return nil
+	})
+}
+
+// GetAppliedConfigurations returns all configurations that have been marked as applied, ordered newest -> oldest.
+func (st *DBstore) GetAppliedConfigurations(ctx context.Context, query *models.GetAppliedConfigurationsQuery) error {
+	return st.SQLStore.WithDbSession(ctx, func(sess *db.Session) error {
+		cfgs := []*models.AlertConfiguration{}
+		err := sess.Table("alert_configuration_history").
+			Desc("id").
+			Where("org_id = ? AND applied_at != 0", query.OrgID).
```

Comment:
Is `applied_at != 0` correct? Since it's nullable, won't this return rows with a null `applied_at`?

Reply:
This wouldn't be covered by the current tests, too - we should add a test for the negative case here - that includes a non-applied configuration, and asserts that it's not returned.

Reply:
> Is applied_at != 0 correct? Since it's nullable, won't this return rows with a null applied_at?

Since the type in the struct is `int64` (not a pointer), the value always defaults to `0`.
Do you know in which case it could be saved as null if we're not using a pointer? Or do you think there's a better option? E.g, making the field not nullable and deal with zero-values.

> This wouldn't be covered by the current tests, too - we should add a test for the negative case here - that includes a non-applied configuration, and asserts that it's not returned.

Totally agree. I was planning on doing this in my next PR. This PR focuses on the write path, the next one will focus on the logic for retrieving records and API-specific things.

Reply:
I think the default of 0 is dangerous:
1. It's a nullable int, can't we set `DEFAULT NULL`?
2. This is a unix timestamp, `0` can be interpreted as `1970-01-01T00:00:00Z`.

If we disagree with the above points and would rather still keep 0 as the default "empty" representation, let's make the column `nullable: false` instead. Let's settle on one default representation here, and not have two different ones.

Reply:
You're right! 0 could be interpreted as Jan 1st 1970, that's why I treat this value as an `int64` everywhere (both struct and DB). We use `int64` for other timestamps, e.g. `created_at`.

> It's a nullable int, can't we set DEFAULT NULL and make the default whatever we like?

Could you explain this a bit more in depth?

I'm totally open to changing this default value, I agree it could be dangerous. However, I currently can't see how having a pointer to a value is any less dangerous, and I think I'm not getting which other values could be used instead.

I agree with making the column not nullable if we decide to stick with ints.

Reply:
Agree with @alexweav here, either (nullable+null) or (not null + 0). 

In addition, why shouldn't we use `time.Time` instead of `int64`? Seems even clearer to me.

Reply:
After talking with @JacobsonMT and @alexweav, consensus is using an `int64` and using not null + 0 (that's how we're currently handling timestamps [in other places](https://github.com/grafana/grafana/blob/c6f4eb4282192bac212f5987960bc29d76261ee3/pkg/services/ngalert/models/alertmanager.go#L12)).